### PR TITLE
Add typing module to IO

### PR DESCRIPTION
**Pull request by nstarman**
Describes PathLike and two flavors of file-like

- [ ] By checking this box, the PR author has requested that maintainers do **NOT** use the "Squash and Merge" button. Maintainers should respect this when possible; however, the final decision is at the discretion of the maintainer that merges the PR.

### CODE REVIEW

File: astropy/io/typing.py [hunk 1]
```diff
@@ -0,0 +1,53 @@
...
+    checking with Protocols works.
+    """
+
+    def read(self) -> _T_co:
+        ...
+
+
+@runtime_checkable
+class WriteableFileLike(Protocol[_T_contra]):
+    """Protocol for file-like objects that are writeable.
```
Comment (nstarman) [inline astropy/io/typing.py:44]:
```suggestion
    """A file-like object that supports writing with a method ``write``.
```

File: astropy/io/typing.py [hunk 2]
```diff
@@ -0,0 +1,53 @@
...
+
+__all__ = ["PathLike", "ReadableFileLike", "WriteableFileLike"]
+
+import os
+from typing import TYPE_CHECKING, Protocol, TypeVar, Union, runtime_checkable
+
+if TYPE_CHECKING:
+    from typing import TypeAlias
+
+_T_co = TypeVar("_T_co", covariant=True)
```
Comment (mhvk) [inline astropy/io/typing.py:18]: This part I really have no idea what is being done...

**Review by mhvk — APPROVED** [review]
I cannot really judge correctness here, but it certainly looks good. Would it make sense to annotate some actual pieces of code? E.g., `fits.open` (`hdu.hudlist.fitsopen`)? But also fine to do things in follow-up (feeling free to still change this; nice to get it in well before 6.1 so there is time to iterate if needed).

**Comment by nstarman** [comment]
> Would it make sense to annotate some actual pieces of code? E.g., fits.open (hdu.hudlist.fitsopen)?

See #15917!

File: docs/io/typing.rst
Comment (eerovaher) [inline docs/io/typing.rst:1]: We have just started adding type annotations to `astropy`, so we might want to make big changes to them. If we are going to have public documentation about typing then there should be a warning somewhere that makes it clear that typing support in `astropy` is experimental and subject to change without notice.

File: astropy/io/typing.py [hunk 3]
```diff
@@ -0,0 +1,53 @@
...
+_T_contra = TypeVar("_T_contra", contravariant=True)
+
+PathLike: TypeAlias = Union[str, bytes, os.PathLike]
```
[the remark this answers is not in the document]
Comment (nstarman) [reply astropy/io/typing.py:21]: It's quite annoying that Python made `os.PathLike` given that their own definition of path-like includes `str` and `bytes`.

[the remark this answers is not in the document]
Comment (eerovaher) [reply astropy/io/typing.py:21]: We can reconsider this if it ever becomes a problem.

**Pull request merged**